source/module_hamilt_lcao/hamilt_lcaodft/operator_lcao/td_nonlocal_lcao.cpp

```diff
@@ -0,0 +1,399 @@
+
+    const Parallel_Orbitals* paraV = this->hR_tmp->get_atom_pair(0).get_paraV();
+    const int npol = this->ucell->get_npol();
+    // 1. calculate <psi|beta> for each pair of atoms
+#ifdef _OPENMP
+#pragma omp parallel
+{
+    std::unordered_set<int> atom_row_list;
+    #pragma omp for
+    for (int iat0 = 0; iat0 < this->ucell->nat; iat0++)
+    {
+        atom_row_list.insert(iat0);
+    }
+#endif
+    for (int iat0 = 0; iat0 < this->ucell->nat; iat0++)
+    {
+        auto tau0 = ucell->get_tau(iat0);
+        int T0, I0;
+        ucell->iat2iait(iat0, &I0, &T0);
+        AdjacentAtomInfo& adjs = this->adjs_all[iat0];
+
+        std::vector<std::unordered_map<int, std::vector<std::complex<double>>>> nlm_tot;
+        nlm_tot.resize(adjs.adj_num + 1);
+
+        for (int ad = 0; ad < adjs.adj_num + 1; ++ad)
+        {
+            const int T1 = adjs.ntype[ad];
+            const int I1 = adjs.natom[ad];
+            const int iat1 = ucell->itia2iat(T1, I1);
+            const ModuleBase::Vector3<double>& tau1 = adjs.adjacent_tau[ad];
+            const Atom* atom1 = &ucell->atoms[T1];
+
+            const ORB_gen_tables& uot = ORB_gen_tables::get_const_instance();
+            const LCAO_Orbitals& orb = LCAO_Orbitals::get_const_instance();
+            auto all_indexes = paraV->get_indexes_row(iat1);
+#ifdef _OPENMP
+            if(atom_row_list.find(iat1) == atom_row_list.end())
+            {
+                all_indexes.clear();
+            }
+#endif
+            auto col_indexes = paraV->get_indexes_col(iat1);
+            // insert col_indexes into all_indexes to get universal set with no repeat elements
+            all_indexes.insert(all_indexes.end(), col_indexes.begin(), col_indexes.end());
+            std::sort(all_indexes.begin(), all_indexes.end());
+            all_indexes.erase(std::unique(all_indexes.begin(), all_indexes.end()), all_indexes.end());
+            for (int iw1l = 0; iw1l < all_indexes.size(); iw1l += npol)
+            {
+                const int iw1 = all_indexes[iw1l] / npol;
+                std::vector<std::vector<std::complex<double>>> nlm;
+                // nlm is a vector of vectors, but size of outer vector is only 1 here
+                // If we are calculating force, we need also to store the gradient
+                // and size of outer vector is then 4
+                // inner loop : all projectors (L0,M0)
+#ifdef USE_NEW_TWO_CENTER
+                // snap_psibeta_half_tddft() are used to calculate <psi|exp(iAr)|beta> no matter whether to use new two-center or not for now.
+                int L1 = atom1->iw2l[ iw1 ];
+                int N1 = atom1->iw2n[ iw1 ];
+                int m1 = atom1->iw2m[ iw1 ];
+
+                // convert m (0,1,...2l) to M (-l, -l+1, ..., l-1, l)
+                int M1 = (m1 % 2 == 0) ? -m1/2 : (m1+1)/2;
+
+                ModuleBase::Vector3<double> dtau = tau0 - tau1;
+                uot.snap_psibeta_half_tddft(orb,
```

Comment:
@jinzx10 do you have good idea for this two-center integral interface?

Reply:
> @jinzx10 do you have good idea for this two-center integral interface?

If I understand correctly, this function is calculating integrals like

$$
\int \phi_{\mu}(\mathbf{r}) e^{i\mathbf{q}\cdot\mathbf{r}} \phi_{\nu}(\mathbf{r}-\mathbf{R}) \mathrm{d}\mathbf{r}
$$

where the $\phi$'s have the form of a pseudo-atomic wavefunction ([radial] x [spherical harmonics]). Sadly, I think this type of integral does not fall into the category of two-center integrals because the $e^{i\mathbf{q}\cdot\mathbf{r}} $ part breaks the whole thing: its expansion contains an infinite number of pseudo-atomic wave function components and is thus inappropriate to use two-center integral techniques.  And indeed, the present implementation uses a grid integral with the Lebedev grid.